### DISCUSSION TITLE
Welcome to Tempurpedic_Control Discussions!

### DISCUSSION THREAD

**Discussion by Doug-Wyman** [Doug-Wyman/Tempurpedic_Control#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by ky13d3nt0n** [comment]
Working on something similar but for Apple shortcuts, I've been trying to track down all of codes for controlling these darn beds, but having a tough time until I came across your repo. Anyway you could share the 30 massage codes?

I would help with the HA setup but I unfortunately don't know Python.

**Reply by Doug-Wyman** [reply]
The *.py files are plain text and in each of the bed commands is a line "message = "... and the hex code for each command.
I should probably add a table of the hex codes as a file in this repository.

**Reply by Doug-Wyman** [reply]
@ky13d3nt0n the file of hex codes is available on the repository.  Thanks for the idea

**Reply by ky13d3nt0n** [reply]
@Doug-Wyman - yep, was able to track down all the codes you currently have, but thanks for the file! In your readme, you also mentioned that there are are 30 individual codes for the massage?

"Tempurpedic has codes for 0 to 10 levels of vibration for each of the three zones (upper body, hips and legs). If I can find an easy way of implementing these 30 levels I will do so."

Curious if you already have these code jotted down somewhere as they are the last piece of functionality I'm missing. May even look into how I can translate this into perhaps a node app for Home Assistant.

**Reply by Doug-Wyman** [reply]
Ok, I have the codes.  Sort of... 👍🏻 
The last byte is (I am fairly sure) some sort of checksum to insure packet correctness
The following is the head range .  
Head vib range
33 05 32 03 94 85 00 18  09 
33 05 32 03 94 85 00 30  21 
33 05 32 03 94 85 00 48  59 
33 05 32 03 94 85 00 60  71
33 05 32 03 94 85 00 78  69   
33 05 32 03 94 85 00 90  81   
33 05 32 03 94 85 00 a8  b9    
33 05 32 03 94 85 00 c0  d1   
33 05 32 03 94 85 00 d8  c9   
33 05 32 03 94 85 00 f0  e1      

The third from last byte is the designator for head, torso or legs  so the  
so the Torso would be:  (editing to add some I calculated)  
torso vib range
33 05 32 03 94 85 01 18  08
33 05 32 03 94 85 01 30 20
33 05 32 03 94 85 01 48  58
33 05 32 03 94 85 01 60  70
33 05 32 03 94 85 01 78  68
33 05 32 03 94 85 01 90  80
33 05 32 03 94 85 01 a8  b8
33 05 32 03 94 85 01 c0  d0  
33 05 32 03 94 85 01 d8  c8
33 05 32 03 94 85 01 f0  e0

and the legs the:
33 05 32 03 94 85 02 18 0b
33 05 32 03 94 85 02 30 23
33 05 32 03 94 85 02 48 5b
33 05 32 03 94 85 02 60 73
33 05 32 03 94 85 02 60 73
33 05 32 03 94 85 02 78 6b
33 05 32 03 94 85 02 90 83
33 05 32 03 94 85 02 a8  bb
33 05 32 03 94 85 02 c0  d3  
33 05 32 03 94 85 02 d8  cb
33 05 32 03 94 85 02 f0  e3

**Reply by ky13d3nt0n** [reply]
Thank you so much! I was able to figure them all out! CRC definitely something I wasn't familiar with but was able to figure out how they were being calculated.

There was also a level 0 for each of the massage areas that I was able to sorta guess so you can manually turn off each section rather than turning them all off at once with the stop command.

Here's all the commands/codes I have now:
```
commands: {
    /**
     * Flat Position
     */
    flat: { title: 'Flat Position', command: '33 05 32 0A 94 5C 04 00 CC' },

    /**
     * Memory Position #1 - #4
     */
    memory1: { title: 'Memory Position 1', command: '33 05 32 03 94 5C 00 00 C8' },
    memory2: { title: 'Memory Position 2', command: '33 05 32 03 94 5C 01 00 C9' },
    memory3: { title: 'Memory Position 3', command: '33 05 32 03 94 5C 02 00 CA' },
    memory4: { title: 'Memory Position 4', command: '33 05 32 03 94 5C 03 00 CB' },

    /**
     * Off
     */
    off: { title: 'Vibration Off', command: '33 05 32 0A 94 86 00 00 12' },

    /**
     * Vibration Setting #1 - #4
     */
    vibration1: { title: 'Vibration Setting 1', command: '33 05 32 03 94 8D 00 78 61' },
    vibration2: { title: 'Vibration Setting 2', command: '33 05 32 03 94 8D 01 78 60' },
    vibration3: { title: 'Vibration Setting 3', command: '33 05 32 03 94 8D 02 78 63' },
    vibration4: { title: 'Vibration Setting 4', command: '33 05 32 03 94 8D 03 78 62' },

    /**
     * Lift
     */
    lift: {
      head: {
        up: { title: 'Lift Head', command: '33 05 32 18 94 53 00 05 C2' },
        down: { title: 'Lower Head', command: '33 05 32 18 94 54 00 05 C5' }
      },
      leg: {
        up: { title: 'Lift Leg', command: '33 05 32 18 94 51 01 00 C4' },
        down: { title: 'Lower Leg', command: '33 05 32 18 94 52 01 00 C7' }
      }
    },

    /**
     * Massage
     */
    massage: {
      head: {
        lvl0: { title: 'Level 0 Massage', command: '33 05 32 03 94 85 00 00 11' },
        lvl1: { title: 'Level 1 Massage', command: '33 05 32 03 94 85 00 18 09' },
        lvl2: { title: 'Level 2 Massage', command: '33 05 32 03 94 85 00 30 21' },
        lvl3: { title: 'Level 3 Massage', command: '33 05 32 03 94 85 00 48 59' },
        lvl4: { title: 'Level 4 Massage', command: '33 05 32 03 94 85 00 60 71' },
        lvl5: { title: 'Level 5 Massage', command: '33 05 32 03 94 85 00 78 69' },
        lvl6: { title: 'Level 6 Massage', command: '33 05 32 03 94 85 00 90 81' },
        lvl7: { title: 'Level 7 Massage', command: '33 05 32 03 94 85 00 A8 B9' },
        lvl8: { title: 'Level 8 Massage', command: '33 05 32 03 94 85 00 C0 D1' },
        lvl9: { title: 'Level 9 Massage', command: '33 05 32 03 94 85 00 D8 C9' },
        lvl10: { title: 'Level 10 Massage', command: '33 05 32 03 94 85 00 F0 E1' }
      },
      lumbar: {
        lvl0: { title: 'Level 0 Massage', command: '33 05 32 03 94 85 01 00 10' },
        lvl1: { title: 'Level 1 Massage', command: '33 05 32 03 94 85 01 18 08' }, // CRC is last byte calced from last 8 digits
        lvl2: { title: 'Level 2 Massage', command: '33 05 32 03 94 85 01 30 20' },
        lvl3: { title: 'Level 3 Massage', command: '33 05 32 03 94 85 01 48 58' },
        lvl4: { title: 'Level 4 Massage', command: '33 05 32 03 94 85 01 60 70' },
        lvl5: { title: 'Level 5 Massage', command: '33 05 32 03 94 85 01 78 68' },
        lvl6: { title: 'Level 6 Massage', command: '33 05 32 03 94 85 01 90 80' },
        lvl7: { title: 'Level 7 Massage', command: '33 05 32 03 94 85 01 A8 B8' },
        lvl8: { title: 'Level 8 Massage', command: '33 05 32 03 94 85 01 C0 D0' },
        lvl9: { title: 'Level 9 Massage', command: '33 05 32 03 94 85 01 D8 C8' },
        lvl10: { title: 'Level 10 Massage', command: '33 05 32 03 94 85 01 F0 E0' }
      },
      leg: {
        lvl0: { title: 'Level 0 Massage', command: '33 05 32 03 94 85 02 00 13' },
        lvl1: { title: 'Level 1 Massage', command: '33 05 32 03 94 85 02 18 0B' },
        lvl2: { title: 'Level 2 Massage', command: '33 05 32 03 94 85 02 30 23' },
        lvl3: { title: 'Level 3 Massage', command: '33 05 32 03 94 85 02 48 5B' },
        lvl4: { title: 'Level 4 Massage', command: '33 05 32 03 94 85 02 60 73' },
        lvl5: { title: 'Level 5 Massage', command: '33 05 32 03 94 85 02 78 6B' },
        lvl6: { title: 'Level 6 Massage', command: '33 05 32 03 94 85 02 90 83' },
        lvl7: { title: 'Level 7 Massage', command: '33 05 32 03 94 85 02 A8 BB' },
        lvl8: { title: 'Level 8 Massage', command: '33 05 32 03 94 85 02 C0 D3' },
        lvl9: { title: 'Level 9 Massage', command: '33 05 32 03 94 85 02 D8 CB' },
        lvl10: { title: 'Level 10 Massage', command: '33 05 32 03 94 85 02 F0 E3' }
      }
    }
  }
```

**Reply by ky13d3nt0n** [reply]
Oh shoot! I didn't even see you posted the updated checksums! Either way learned something new!

**Reply by Doug-Wyman** [reply]
That is what life is all about, LEARNING Turning 80 in Nov and won't stop learning till at least 100.

**Comment by ky13d3nt0n** [comment]
Hey @Doug-Wyman wanted to share an update!

I tried so many times to write this functionality in Python, but I'm a JS developer at my core so it was rough. Instead, I opted to expose the functionality of my Ergo Base via ESPHome. It's a pretty simple setup, the board costs about $2-3 bucks and if you have the ESPHome add-on you can easily flash the board.

I've created a gist file example and wanted to share what it looks like in HomeAssistant.

Gist: https://gist.github.com/ky13d3nt0n/5f0cb98d3939a895a8c30e41e1a56c24

![screencapture-home-assistant-dev-denton-app-config-devices-device-eb31ab427b5bf8fcc1dbdb4ebdd643c5-2023-08-30-20_41_08](https://github.com/Doug-Wyman/Tempurpedic_Control/assets/52417040/be8e5611-fc9d-4b5b-8566-936f90c4f1b8)

Still some kinks to work out and it probably has a bug or two, but so far I'm pretty happy with the setup.

Also just realized it's been just over 1 year since you shared the hex codes, just wanted to say thanks again for all of your help!

**Reply by Doug-Wyman** [reply]
This is pretty awesome.  It makes me feel good that it is evolving.  My motto has for a long time been, "Never stop learning. Never stop evolving."  Keep it up and thanks for the update!

**Comment by syniast** [comment]
Am I supposed to set this up in the add-ons store under custom repo?  I get this error trying to add this git as a custom repo.   https://github.com/Doug-Wyman/Tempurpedic_Control is not a valid add-on repository

**Reply by Doug-Wyman** [reply]
No.  There are .yaml files that, when added to the configuration.yaml create the entities to call the commands.  I even included a dashboard .yaml file.  If you are unfamilier with Home Assistant, be sure to do a back-up before mucking around the config.yaml but that is also the beauty of Home Assistant.  If you have a bed and want to control it Home Assistant is flexible enough to do so.  Have at it and kindly please drop this old man a reply on here when you get it up or before <grin>
